File: watchtower/internal/model/semconv.go
package model

// Semconv attribute names shared across the wire: the Python harness
// writes these, the Go verifiers read them. The gen_ai.* names follow the
// OTel GenAI semantic conventions (operation, system, model, token
// counts); tool/agent attributes align with the gen-ai agent semconv
// direction; watchtower.* are our own additions (structured-output
// contract checking).
//
// Keeping every attribute name in one place is what makes the pipeline
// machine-checkable: a typo here breaks the contract between harness and
// verifiers, and the compiler catches it on the Go side.
const (
	// GenAI operation (e.g. "chat"), system (provider), and model ids.
	GenAIOperationName  = "gen_ai.operation.name"
	GenAISystem         = "gen_ai.system"
	GenAIRequestModel   = "gen_ai.request.model"
	GenAIResponseModel  = "gen_ai.response.model"
	GenAIInputTokens    = "gen_ai.usage.input_tokens"
	GenAIOutputTokens   = "gen_ai.usage.output_tokens"
	GenAICompletionTime = "gen_ai.usage.completion_time" // ours: ms

	// Tool-call spans.
	ToolName      = "tool.name"
	ToolCallID    = "tool.call.id"
	ToolResultOK  = "tool.result.ok" // ours: "true"/"false"
	ToolResultMsg = "tool.result.message"

	// Agent spans: which agent owns this span, and what step index it is.
	AgentName = "agent.name"
	AgentID   = "agent.id"
	StepIndex = "agent.step" // ours: 0-based step number within the agent

	// Watchtower-specific: structured-output contract checking. The
	// harness stamps the contract name it is supposed to honor and the
	// raw model output; the schema verifier parses and checks it.
	WatchtowerContract = "watchtower.contract"
	WatchtowerOutput   = "watchtower.output"
)

File: harness/semconv.py
"""Semconv attribute names — the vocabulary both sides of the wire use.

These mirror the Go constants in watchtower/internal/model/semconv.go.
The wire contract is these strings; the Go side compiles against them,
so the Python side must treat them as frozen. A typo here silently
disables a verifier, so keep this file in lockstep with the Go one.
"""

GEN_AI_OPERATION_NAME = "gen_ai.operation.name"
GEN_AI_SYSTEM = "gen_ai.system"
GEN_AI_REQUEST_MODEL = "gen_ai.request.model"
GEN_AI_RESPONSE_MODEL = "gen_ai.response.model"
GEN_AI_INPUT_TOKENS = "gen_ai.usage.input_tokens"
GEN_AI_OUTPUT_TOKENS = "gen_ai.usage.output_tokens"

TOOL_NAME = "tool.name"
TOOL_CALL_ID = "tool.call.id"
TOOL_RESULT_OK = "tool.result.ok"
TOOL_RESULT_MSG = "tool.result.message"
WATCHTOWER_TOOL_ARGS = "watchtower.tool.args"

AGENT_NAME = "agent.name"
AGENT_ID = "agent.id"
STEP_INDEX = "agent.step"
WATCHTOWER_COMPLETED = "watchtower.completed"
WATCHTOWER_FINAL = "watchtower.final"
WATCHTOWER_OUTCOME = "watchtower.outcome"
WATCHTOWER_TOOL_CALL_IDS = "watchtower.tool.call.ids"

WATCHTOWER_CONTRACT = "watchtower.contract"
WATCHTOWER_OUTPUT = "watchtower.output"
